[FIX] SheetPlugin: Prevent deletion of all non-frozen headers

dhrp-odoo opened the pull request:
## Description:

This PR contains two commits that tackle the following issues:

***Commit 1:***

Issue: Deleting rows/columns, except the frozen ones, would break the spreadsheet. Hiding some rows and then deleting all rows, except the frozen ones, or using a data filter to hide some rows and then deleting all rows, would also cause issues.

Solution: Implemented a validation mechanism in the `SheetPlugin.` The `allowDispatch` function now checks for attempts to delete all rows or columns when frozen panes are present. If such actions are detected, the corresponding command will be rejected.

***Commit 2:***

Issue: The `searchHeaderIndex` function returned `-1` when trying to hide all rows/cols, leading to a broken sheet. This occurred because the `adjustViewportZoneY` function used this value as top.

Solution: Modified the `searchHeaderIndex` function to handle `top` and `left` `-1` cases correctly. Now, when trying to hide all rows or columns, the function reassigns the value of `boundaries.top` and `boundaries.left` respectively, resolving the issue.

Task: [3414127](https://www.odoo.com/web#id=3414127&cids=2&menu_id=4720&action=333&active_id=2328&model=project.task&view_type=form)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_lt("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/o-spreadsheet/pull/2673).

hokolomopo commented on src/plugins/core/sheet.ts:
```diff
@@ -233,6 +256,39 @@ export class SheetPlugin extends CorePlugin<SheetState> implements SheetState {
+
+    const hiddenColsRowsGroups: Set<HeaderIndex> = new Set(
+      dimension === "ROW"
+        ? this.getters.getHiddenRowsGroups(sheetId).flat()
```
Just scrolled through the pr really fast, and this won't work because, as I noted in the pad, the headers can be hidden by UI features (data filters). I'm not 100% sur on the best way to proceed, but this definitely shouldn'tbe handled in core (or maybe as a temporary stopgag measure).

dhrp-odoo replied:
Hi, I'm aware of this issue. Currently this code fixes the issue with delete and hide of rows/cols in normal cases. I'm trying to figure out a way to fix data filters, but it seems rather tricky.

Second point you mentioned that I should not include it in core, well freeze is originally controlled in sheet.ts, so it made more sense to modify the allowDispatch there. I'll try to look for more alternatives

rrahir reviewed (changes requested):
Several comments:
1.this requires at least a rebase - we have updated some APIs
2. Looking at the video in the task I feeel like you go beyond the scope that we wanted to cover at first which  is - you cannot delete all the rows/columns of the unfrozen parts of a spreadsheet. I don't understand how this is related to their visibility,

dhrp-odoo commented on src/plugins/ui/header_visibility_ui.ts:
```diff
@@ -12,6 +21,66 @@ export class HeaderVisibilityUIPlugin extends UIPlugin {
+        }
+
+        if (this.isAllHeaderIndices(cmd.sheetId, cmd.dimension, cmd.elements)) {
+          this.ui.notifyUI({
```
Hi, I am aware that we have updated the API to catch user errors. However, since it was in the master branch, I thought I would update my code there.

Nevertheless, I completely agree with you and will handle this using the example you provided.

dhrp-odoo commented on src/helpers/internal_viewport.ts:
```diff
@@ -365,6 +365,9 @@ export class InternalViewport {
       this.searchHeaderIndex("COL", this.viewportWidth, this.left)
     );
+    if (this.left === -1) {
+      this.left = this.getters.getPaneDivisions(sheetId).xSplit;
```
![image](https://github.com/odoo/o-spreadsheet/assets/121081869/086ae83e-5f3e-4003-bd98-9bfde9f8514e)

If we attempt to filter all rows using a data filter, in this case, the `searchHeaderIndex` function returns -1, resulting in a broken sheet.

dhrp-odoo commented on src/plugins/ui/header_visibility_ui.ts:
```diff
@@ -12,6 +21,66 @@ export class HeaderVisibilityUIPlugin extends UIPlugin {
 
+  allowDispatch(cmd: Command): CommandResult {
+    switch (cmd.type) {
+      case "REMOVE_COLUMNS_ROWS":
```
I tried to implement the fix in core earlier, but as suggested by ADRM
Here: https://github.com/odoo/o-spreadsheet/pull/2673#discussion_r1262465671
Discord: https://discord.com/channels/678381219515465750/687337979743830044/1131556525123571732

This approach makes sense since we won't be able to access the rows hidden by the UI data filter, which necessitates keeping our code within the UI.

If we don’t go beyond the issues written in specs of this task, then I believe we can skip the data filter validation part, and can ONLY handle the removal of columns/rows (via the core).

Since we have a similar file in core and (this) as an extended part in UI, I thought it would be the best place to drop CORE commands through allowDispatch. Although, in future versions (16.3 & above) this won’t work either. So, going to core would be the only choice but then again data filter has to be handled separately in UI.

hokolomopo reviewed (commented):
Hey ! 

I think your commit on the internal viewport is mostly good, just take into account @rrahir comment (replace `this.getters.getPaneDivisions(sheetId).xSplit` with `this.boundaries.left` maybe? I'm not 100% sure, I'm not too familiar with the viewport's code).

The commit on HeaderVisibilityUIPlugin is just plain wrong tho. You cannot and should not user allowDispatch on a _core_ command in an _UI_ plugin. This either break the collaborative if a command was accepted by an user and not the other, or won't work at all if the remote command don't go through the allowDispatch.

The good news is, you don't really need to do this anymore ! Your other commit make it so the viewport won't break if all the rows that are not in the frozen pane are hidden/deleted. So you could leave it alone.

But I think you're right and it's nice to prevent the user from deleting/hiding all the non-frozen rows. But this is impossible to include the data filters in all this. We have to keep everything in core, and ignore the data filters (which, again, do'nt break the sheet anymore with your other commit).

What i'd do is to update the getter `canRemoveHeaders` with what you've done, so we cannot hide/remove all the rows below the frozen pane. The getter is used in the `allowDispatch`, and is used to disable the menu items. So you do'nt have to notify the user, since the menu items to make the wrong action aren't there in the first place. 

If you discussed the error message with Francois, and he'd rather have an error message than the disabled menu items, you should write an `interactiveDeleteRows` method instead of using a notifyUI in a  plugin (look at `interactiveAddMerge` for an example).

Let me know if something is unclear or you have any questions.

dhrp-odoo commented:
Hello @hokolomopo,

Indeed, RAR's observation was correct. The solution for the Internal Viewport was overly broad, I've implemented the changes as per suggestion. 

Regarding the second commit, I've introduced an `interactiveRemoveColumnsRows` method. This method rejects the command, generating a result that can be managed through an interactive call. The corresponding command is handled within the core's `SheetPlugins allowDispatch`.

Additionally, I believe that we no longer need to be concerned about hidden rows. Please let me know if you have any potential edge cases in mind.

Please review the changes again at your earliest convenience.

Thank you

hokolomopo commented on src/plugins/core/sheet.ts:
```diff
@@ -233,6 +235,36 @@ export class SheetPlugin extends CorePlugin<SheetState> implements SheetState {
+   * Checks if all non-frozen header indices are present in the provided elements of selected rows/columns.
+   * This validation ensures that all rows or columns cannot be deleted when frozen panes exist.
+   */
+  private checkHeaderVisibility(
```
function name is wrong, this has nothing to do with visibility

hokolomopo commented on src/plugins/core/sheet.ts:
```diff
@@ -233,6 +235,36 @@ export class SheetPlugin extends CorePlugin<SheetState> implements SheetState {
+
+    for (let i = startIndex; i <= endIndex; i++) {
+      if (!elements.includes(i)) {
+        // If the index is not present in the elements of selected rows/columns, then return false.
```
This comment is kinda usless, you juste explain 2 easy to understand lines of code.

hokolomopo commented on src/helpers/ui/sheet_interactive.ts:
```diff
@@ -38,3 +38,24 @@ export function interactiveRenameSheet(env: SpreadsheetChildEnv, sheetId: UID, e
+    elements: elements,
+  });
+  const errorMessage = _lt(
+    `Sorry, it is not possible to delete all non-frozen ${
+      dimension === "ROW" ? "rows" : "columns"
+    }.`
```
translations don't work with backticks. Use standard string with placeholders `%s`

hokolomopo reviewed (commented):
Some details and we'll be good I think

hokolomopo commented:
Also remove the commit to "prevent deletion of all non-frozen headers". Your current title make it looks like we can delete _all_ the headers when there's a frozen pane

hokolomopo commented on src/plugins/core/sheet.ts:
```diff
@@ -233,6 +235,35 @@ export class SheetPlugin extends CorePlugin<SheetState> implements SheetState {
+  ): boolean {
+    const paneDivisions = this.getters.getPaneDivisions(sheetId);
+    const startIndex = dimension === "ROW" ? paneDivisions.ySplit : paneDivisions.xSplit;
+    const endIndex =
+      dimension === "ROW"
+        ? this.getters.getNumberRows(sheetId) - 1
+        : this.getters.getNumberCols(sheetId) - 1;
```
use `getNumberHeaders`

hokolomopo reviewed (commented):
Thinking about it some more, what you propose isn't wrong but it give us a consistency issue: removing all the cols is not possible because the menu item is hidden, but removing all the non-frozen header do'nt hide the menu item but gives an error message. We really should do the same thing every time. 

I'd keep the menu item hidden for everything, since it is what we had before. You should make your function a getter, and call it in the isVisible of the menu items. You can also probably simplify the writing of it by using the helper  `includesAll` (see in `canRemoveHeaders`).  Also rename `canRemoveHeaders` to `checkElementsIncludeAllVisibleHeaders` to be consistent.

dhrp-odoo commented:
> Thinking about it some more, what you propose isn't wrong but it give us a consistency issue: removing all the cols is not possible because the menu item is hidden, but removing all the non-frozen header do'nt hide the menu item but gives an error message. We really should do the same thing every time.
> 
> I'd keep the menu item hidden for everything, since it is what we had before. You should make your function a getter, and call it in the isVisible of the menu items. You can also probably simplify the writing of it by using the helper `includesAll` (see in `canRemoveHeaders`). Also rename `canRemoveHeaders` to `checkElementsIncludeAllVisibleHeaders` to be consistent.

I've made the suggested changes, including adding the test case.

hokolomopo commented on src/plugins/core/header_visibility.ts:
```diff
@@ -97,7 +97,11 @@ export class HeaderVisibilityPlugin extends CorePlugin {
+    elements: HeaderIndex[]
+  ): boolean {
     const visibleHeaders = this.getAllVisibleHeaders(sheetId, dimension);
     return !includesAll(elements, visibleHeaders);
```
```suggestion
    return includesAll(elements, visibleHeaders);
```
remove the `!` and update the calls of this method. With the "!" it does the opposite of wha its name says

hokolomopo commented on src/helpers/ui/sheet_interactive.ts:
```diff
@@ -38,3 +38,23 @@ export function interactiveRenameSheet(env: SpreadsheetChildEnv, sheetId: UID, e
   });
 }
+
+export function interactiveRemoveColumnsRows(
```
Remove the interactive method, sicne we decided to go with hidden menu items instead.

hokolomopo replied:
The topbar also uses the menu item, so the option won't show either :)

dhrp-odoo replied:
Well, In my case the option still appears from the top bar edit.
Maybe my fix was not in the right place, let me check

hokolomopo replied:
Mh yeah we didn't use the exact same menu item in 16.0. Extract the isVisible in a helper `CAN_REMOVE_COLUMNS` then, and put it also in the topbar edit menu item.

dhrp-odoo replied:
I have implemented the suggested changes; It should look better now :)

hokolomopo commented on tests/components/grid_manipulation.test.ts:
```diff
@@ -150,6 +151,20 @@ describe("Context Menu add/remove row/col", () => {
     expect(fixture.querySelector(".o-menu div[data-name='hide_rows']")).toBeNull();
   });
 
+  test("cannot delete all non-frozen rows with contextmenu when frozen panes exist", async () => {
```
The test should be in `menu_items_registry.test.ts`

dhrp-odoo replied:
Since the `cannot delete nor hide all rows with contextmenu` test case is written here, that's why I have included my test case here. [TESTCASE](https://github.com/odoo/o-spreadsheet/blob/34011c3de11b883dbd9ca0494d3ff8e09e5b58c8/tests/components/grid_manipulation.test.ts#L123-L151)

Should we consider adding this to the menu_items_registry.test.ts file?

hokolomopo reviewed (commented):
Hopefully last changes before r+

hokolomopo commented on src/registries/menus/topbar_menu_registry.ts:
```diff
@@ -103,11 +104,13 @@ topbarMenuRegistry
     name: ACTIONS.REMOVE_ROWS_NAME,
     sequence: 80,
     action: ACTIONS.REMOVE_ROWS_ACTION,
+    isVisible: (env: SpreadsheetChildEnv) => CAN_REMOVE_COLUMNS_ROWS("ROW", env),
```
the helper should be at the same place as the other menu item helpers (`lenu_item_action.ts`), and be imported/used like the others  `ACTION.CAN_REMOVE_COLUMNS_ROW`

hokolomopo commented on tests/menu_items_registry.test.ts:
```diff
@@ -301,6 +303,20 @@ describe("Menu Item actions", () => {
+      selectRow(model, 4, "newAnchor");
+      selectRow(model, lastRow, "updateAnchor");
+
+      simulateContextMenu(".o-row-resizer", ROW_5);
+      await nextTick();
+      expect(fixture.querySelector(".o-menu div[data-name='delete_row']")).toBeNull();
```
Change it to use the `getNode(path).isVisible` like the most other tests of this file. The way you did it won't work at the forward port, we optimized this file by not mounting the spreadsheet, disallowing the use of contextClicks to open the menu

dhrp-odoo replied:
Changes made

hokolomopo commented on tests/menu_items_registry.test.ts:
```diff
@@ -363,6 +379,20 @@ describe("Menu Item actions", () => {
+      selectColumn(model, 3, "newAnchor");
+      selectColumn(model, lastColumn, "updateAnchor");
+
+      simulateContextMenu(".o-col-resizer", COLUMN_D);
+      await nextTick();
+      expect(fixture.querySelector(".o-menu div[data-name='delete_column']")).toBeNull();
```
same here

hokolomopo reviewed (commented):
@robodoo r+

thanks for the work 👍

robodoo commented:
@dhrp-odoo @hokolomopo because this PR has multiple commits, I need to know how to merge it:

* `merge` to merge directly, using the PR as merge commit message
* `rebase-merge` to rebase and merge, using the PR as merge commit message
* `rebase-ff` to rebase and fast-forward

hokolomopo commented:
@robodoo rebase-ff

robodoo commented:
Merge method set to rebase and fast-forward.
